### PR TITLE
CGC: Log Run ID after sweep

### LINKED ISSUE #5781: GC: Incremental - Log run ID
Log run ID after a sweep has been completed.

### PR DESCRIPTION
**Pull request by Jonathan-Rosenberg**
### Linked Issue

Closes #5781

---

## Change Description

When completing a sweep, log the run ID as a marker to `_lakefs/retention/gc/run_ids/<run ID>`.
This is needed to implement Incremental GC.

## How was this tested?

Manually on S3 and Azure.

### CODE REVIEW

File: clients/spark/core/src/main/scala/io/treeverse/clients/BulkRemoverFactory.scala [hunk 1]
```diff
@@ -19,7 +13,7 @@ import org.apache.hadoop.conf.Configuration
 import java.net.URI
 import java.nio.charset.Charset
 import java.util.stream.Collectors
-import collection.JavaConverters._
+import scala.collection.JavaConverters._
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/BulkRemoverFactory.scala:16]: Why this change?

File: clients/spark/core/src/main/scala/io/treeverse/clients/BulkRemoverFactory.scala [hunk 2]
```diff
@@ -99,32 +93,24 @@ object BulkRemoverFactory {
...
     }
   }
 
   private class AzureBlobBulkRemover(hc: Configuration, storageNamespace: String)
       extends BulkRemover {
-    val EmptyString = ""
-    val uri = new URI(storageNamespace)
-    val storageAccountUrl = StorageUtils.AzureBlob.uriToStorageAccountUrl(uri)
-    val storageAccountName = StorageUtils.AzureBlob.uriToStorageAccountName(uri)
+    private val EmptyString = ""
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/BulkRemoverFactory.scala:103]: I don't think we ever need this: Scala checks string equality by value so you can just `s != ""`.  (Also the naming is _really_ weird, it looks like a class).

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 3]
```diff
@@ -400,15 +404,17 @@ object GarbageCollector {
...
+                             expiredAddresses,
+                             markID,
+                             region,
+                             storageType,
+                             schema
+                            )
+        logRunId(runID, storageType, storageNSForSdkClient, hc, region)
+        removed
       } else {
         spark.createDataFrame(spark.sparkContext.emptyRDD[Row], schema)
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:419]: In this case we lose the runID, just because we had nothing to sweep _this time_.  Not sure that makes sense.

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 4]
```diff
@@ -434,6 +440,39 @@ object GarbageCollector {
...
+      val storageAccountName = StorageUtils.AzureBlob.uriToStorageAccountName(uri)
+      val containerName = StorageUtils.AzureBlob.uriToContainerName(uri)
+      val blobClient =
+        Azure.getBlobContainerClient(hc, storageAccountUrl, storageAccountName, containerName)
+      val pathArray = uri.getPath.split("/")
+      val key = pathArray.slice(2, pathArray.length).reduce((a1, a2) => a1 + "/" + a2)
+      blobClient
+        .getBlobClient(getRunIDMarkerLocation(runID, key))
+        .upload(runIDMarkerInputStream, 0)
+    }
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:473]: This should be an additional method on some polymorphic object.  Maybe even on BulkRemover?

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 5]
```diff
@@ -635,10 +674,16 @@ object GarbageCollector {
       .toDF(schema.fieldNames: _*)
   }
 
-  private def getMetadataMarkLocation(markId: String, gcAddressesLocation: String) = {
+  private def getMetadataMarkLocation(markId: String, gcAddressesLocation: String): String = {
     s"$gcAddressesLocation/$markId.meta"
   }
 
+  private def getRunIDMarkerLocation(runID: String, storagePrefix: String): String = {
+    var prefix: String = storagePrefix.stripPrefix("/").stripSuffix("/")
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:682]: @itaiad200 is this the right way to do it nowadays?  We've had so much pain around storage namespaces...

File: clients/spark/core/src/main/scala/io/treeverse/clients/StorageClients.scala [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        hc: Configuration,
+        bucket: String,
+        region: String,
+        numRetries: Int
+    ): AmazonS3 =
+      io.treeverse.clients.conditional.S3ClientBuilder.build(hc, bucket, region, numRetries)
+  }
+
+  object Azure {
+
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/StorageClients.scala:28]: Nit: get rid of this blank line,

**Comment by Jonathan-Rosenberg** [comment]
Thanks @arielshaqed! As mentioned in the description, it was tested manually 🙂

File: clients/spark/core/src/main/scala/io/treeverse/clients/BulkRemoverFactory.scala (hunk 2 above)
Comment (Jonathan-Rosenberg) [reply clients/spark/core/src/main/scala/io/treeverse/clients/BulkRemoverFactory.scala:103]: True. This is not a new var, the only thing that changed is the `private` scoping.
I'll take a llok.

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala (hunk 3 above)
Comment (Jonathan-Rosenberg) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:419]: This is not the case.
In this case `shouldSweep == false`, thus it was explicitly requested to not sweep (passed in the `do_sweep=false` flag).
Even if there was nothing to sweep (i.e. the commits file was empty), yet `shouldSweep == true`, the above `if` clause will handle it and the run ID will be logged.

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala (hunk 4 above)
Comment (Jonathan-Rosenberg) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:473]: sounds good, but not as part of the process of the bulk remover (I don't want it to write the ID multiple times by each executed task)

**Comment by Jonathan-Rosenberg** [comment]
@N-o-Z I've added you because there's a minor change that affects UGC as well.

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala (hunk 3 above)
Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:416]:
```scala
// TODO(ariels): Implement "depends-on" between "review code" and "learn to read".
```

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala (hunk 5 above)
Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:639]: Still worried about this one, though!

**Review by arielshaqed — COMMENTED** [review]
Thanks!

Just blocking for "Slasher" @itaiad200's comment whether the this PR trims slashes correctly.

**Review by N-o-Z — COMMENTED** [review]
I'm not sure I have all the context here.
Why are we logging the run ID in a new path in the bucket?

**Comment by N-o-Z** [comment]
@Jonathan-Rosenberg  - I would recommend looking at what we did in UGC.
We save the reports in a path the contains the runID which is also created in a lexicographically descending order.
This way - when listing the reports from the bucket we will always get a list of runID prefixes from latest to oldest.
Consider changing the structure of how you save logs in CGC instead of adding yet an additional path in the bucket that you need to query

**Comment by Jonathan-Rosenberg** [comment]
Thanks @N-o-Z, 
The UGC way was considered but I prefer writing down the sweep-completed run IDs rather than traversing all run IDs and filtering those that actually did a sweep.
The run IDs being used are the same ones that UGC uses, i.e. using the `4000-now` time and the xid package

File: clients/spark/core/src/main/scala/io/treeverse/clients/BulkRemoverFactory.scala (hunk 1 above)
Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/clients/BulkRemoverFactory.scala:14]: ❓

**Review by arielshaqed — APPROVED** [review]
Thanks!  See a small number of nits, though...

File: clients/spark/core/src/main/scala/io/treeverse/clients/BulkRemoverFactory.scala (hunk 1 above)
Comment (Jonathan-Rosenberg) [reply clients/spark/core/src/main/scala/io/treeverse/clients/BulkRemoverFactory.scala:14]: It was an automatic import optimization, but there's no difference- both imports point to the same library
